# The main entry point of your workflow.
# After configuring, running snakemake -n in a clone of this repository should successfully execute a dry-run of the workflow.

# Define global variables

CHROMS = list(range(1, 23))
GENOME_BUILD = [37, 38]


report: "report/workflow.rst"

# Allow users to fix the underlying OS via singularity.
singularity: "docker://continuumio/miniconda3"


rule all:
    input:
        # The first rule should define the default target files
        # Subsequent target rules can be specified below. They should start with all_*.
        # qc analysis
        "results/QC/note.txt",
        # lift-over to GRCh38
        # unmapped variant to GRCh38
        "results/data/lifted_to_GRCh38/mxb-ALL-GRCh38.vcf.unmap",
        # variant annotation
        expand("results/data/variant-annotation/mxb-chr{chrn}-GRCh{build}.vcf.gz", chrn=CHROMS, build=GENOME_BUILD),


include: "rules/qc-analysis.smk"
include: "rules/phase.smk"
include: "rules/get-raw-genomes.smk"
include: "rules/vep.smk" # Variant annotation
include: "rules/lift-over-to-GRCh38.smk"
