# The main entry point of your workflow.
# After configuring, running snakemake -n in a clone of this repository should successfully execute a dry-run of the workflow.

# Define global variables

CHROMS = list(range(1, 23))


report: "report/workflow.rst"

# Allow users to fix the underlying OS via singularity.
singularity: "docker://continuumio/miniconda3"


rule all:
    input:
        # The first rule should define the default target files
        # Subsequent target rules can be specified below. They should start with all_*.
        # qc analysis
        "results/plots/qc/vars_per_genome.png",
        "results/plots/qc/pca_comp1_2_3.png",
        # variant annotation
        expand("results/data/variant-annotation/vep-{chrn}.vcf.gz", chrn=CHROMS)



include: "rules/get-raw-genomes.smk"
include: "rules/accesory-data.smk"
include: "rules/qc-analysis.smk"
include: "rules/get-raw-genomes.smk"
include: "rules/vep.smk" # Variant annotation
